Feature: register new lua function `DrawBuildSquare` to show build squares on custom building commands

aurelienlt opened the pull request:
Registers `Spring.AddBuildSquare(unitDefId, x, y, z, facing [, opts])` and `Spring.RemoveBuildSquare(unitDefId, x, y, z, facing)` which display the usual grid.

Options are

- `unbuildable=false`: show the building as unbuildable (orange) instead of buildable (green) even if there is no detected conflict.
- `defaultColor={0,0,0,0}`: replace all the hardcoded color with this given color (then it will be overriden by next colors)
- `buildableColor={0,0,0,0}`: change the color for buildable (green)
- `unbuildableColor={0,0,0,0}`: change the color for unbuildable (orange)
- `featureColor={0,0,0,0}`: change the color for feature (orange)
- `illegalColor={0,0,0,0}`: change the color for illegal (red)

Mostly forking `CUnitDrawerLegacy::ShowUnitBuildSquare`.

My first MR on that project, so please lemme know best practices.

![Peek 2024-10-20 19-27](https://github.com/user-attachments/assets/60aaa623-b72d-40c7-a9b5-daf4dfb715b8)

Commit:
Feature: register new lua functions `AddBuildSquare` and `RemoveBuildSquare` to show build squares on custom building commands

sprunk commented:
I think drawing should not be doing any validity calculations (add another function that checks overlap, buildability or whatnot) and should draw a single blueprint.

aurelienlt commented:
This was a copy of the behvior of `CUnitDrawerLegacy::ShowUnitBuildSquare`. 

So you think that we should have a function returning the state of every square associated with a building position, and another function that displays each of this square ?

How would this change help developping scripts that use the building square feature ?

aurelienlt commented:
OK that's quite clear now. I'll work on improving that.

aurelienlt commented:
@sprunk OK I've rewritten the MR to feature a more simple and direct function. Also added some options to allow to edit the colors, I'm thinking about templates which have a specific color for buildings not buildable by the current builder. Lemme know if it's overkill.

sprunk commented:
1) It should accept `unitDefID, x, z, facing` and not `buildCommand`. The interface is more generic than just build commands (e.g. you can draw some sort of "reinforcements will spawn here, clear the way" markers).

2) Native GUI colors are generally defined via "cmdColors", add color customisation there.

3) `bool unbuildable` looks good.

sprunk commented:
@GoogleFrog you originally came up with the suggestion to do it like this so maybe you have some thoughts on API design as well.

aurelienlt commented:
I used commands because it avoided declaring the type buildInfo in lua (command is bascally `{-unitDefID, {x, y, x, facing}, opts}`). But I guess simplifying the type could have some usage.

I also thought about the need of displaying arbitrary squares, but there a caveat to that: the buildability and the height of a square depends on the type of building (some are on or under water, walls can be built on any slope ...). So it would be a different interface anyways that needs to define its own height for each square.

I didn't know about `cmdColors`, better place to define those hardcoded colors. Not sure if we can use it for opts' colors, because the key for custom colors seems to be a commandId. So that would require to register a comand ids `CMD_CANNOT_BUID_HER_FOR_DIFFERENT_REASON`, `CMD_BUILD_HERE_IS_GOOD_IDEA` ot `CMD_PLEASE_CLEAR_THIS_AREA` just to use some specific colors. I think directly passing colors is more simple to use.

I see some LUA calls like `SetLosViewColors` or `SetAtmosphere` to configure colors, so another options could be to add a call to configure default square colors and add arbitrary color labels for specific square displayed.

aurelienlt commented:
I like the idea to draw arbitrary zones to show that a team member plans to use it, or that a zone should be defended. But visually I think it would make more sense to draw it as a polygon with player color borders and lighter color fill. some extension of the map draw, but as a filled shape instead of a line.

sprunk commented:
> I used commands because it avoided declaring the type buildInfo in lua

Well `unitDefID, x, z, facing` is just four numbers so there aren't any issues with types either.

> Not sure if we can use it for opts' colors, because the key for custom colors seems to be a commandId. So that would require to register a comand ids CMD_CANNOT_BUID_HER_FOR_DIFFERENT_REASON, CMD_BUILD_HERE_IS_GOOD_IDEA ot CMD_PLEASE_CLEAR_THIS_AREA just to use some specific colors.

You can just add an arbitrary color entry there. Like here it's using some "restart" color directly which isn't associated with any command.
https://github.com/beyond-all-reason/spring/blob/24efb38cd0be065e67eb9ae6e30721bbdeb6b3b4/rts/Game/SelectedUnitsHandler.cpp#L913

> I see some LUA calls like SetLosViewColors or SetAtmosphere to configure colors, so another options could be to add a call to configure default square colors and add arbitrary color labels for specific square displayed.

There is a Lua call to reload cmdcolors. It has a pretty bad interface that I'm unhappy about but that is outside the scope of this PR.

> the buildability and the height of a square depends on the type of building (some are on or under water, walls can be built on any slope ...). So it would be a different interface anyways that needs to define its own height for each square.

I'm not asking for completely arbitrary squares, just the ability to draw the existing grid for specific buildings. You'd always have to pass a specific unitDefID and then it would draw squares as appropriate for that type. Pass the ID of a floating/underwater building and it will draw squares on/under water accordingly. Pass a wall's ID and it will draw OK for any slope, etc.

aurelienlt commented:
OK I've replaced all the hardcoded colors with colors from `cmdColors` and updated the API to take `unitDefId, x, y, z, facing`. About all those `xxxColor` options, we still keep them or I get rid of them ?

aurelienlt commented:
Oh also I wanted to put those `ParseBuildInfo` functions in `LuaUtils` but it fails to compile. I assum because `LuaUtils` is imported somehwere that doesn't include the code for `BuildInfo`. I guess it can be fixed by a `#if ...` but I have no idea what to check for.

```
./spring/rts/Lua/LuaUtils.cpp:1820: error: undefined reference to 'BuildInfo::BuildInfo(int, float3 const&, int)'
./spring/rts/Lua/LuaUtils.cpp:1863: error: undefined reference to 'BuildInfo::BuildInfo(int, float3 const&, int)'
/usr/include/c++/13/bits/stl_construct.h:97: error: undefined reference to 'BuildInfo::BuildInfo()'
/usr/include/c++/13/bits/stl_construct.h:97: error: undefined reference to 'BuildInfo::BuildInfo()'
```

GoogleFrog commented:
> @GoogleFrog you originally came up with the suggestion to do it like this so maybe you have some thoughts on API design as well.

The discussion looks good. I am not sure what `Spring.DrawBuildSquareArray` is for though.

sprunk commented on rts/Lua/LuaUnsyncedCtrl.cpp:
```diff
@@ -3382,6 +3384,226 @@ int LuaUnsyncedCtrl::SetBuildFacing(lua_State* L)
+	const char* caller,
+	int idx
+) {
+	for (int i=0; i<5; i++) {
+		if (!lua_isnumber(L, idx+i)) {
+			luaL_error(L, "%s(): bad build info type", caller);
+			return false;
+		}
+	}
+
```
Use `luaL_checkinteger` (or float etc. as appropriate) which has a built-in type check, then get rid of all that explicit type check bloat.

aurelienlt replied:
done

sprunk commented on rts/Game/UI/CommandColors.cpp:
```diff
@@ -77,10 +77,17 @@ CCommandColors::CCommandColors()
+	SETUP_COLOR(unbuildableSquare,   0.9f, 0.8f, 0.0f, 0.7f);
+	SETUP_COLOR(featureSquare,       0.9f, 0.8f, 0.0f, 0.7f);
+	SETUP_COLOR(illegalSquare,       0.9f, 0.0f, 0.0f, 0.7f);
+	SETUP_COLOR(undergroundStart,    0.0f, 0.0f, 1.0f, 0.5f);
+	SETUP_COLOR(undergroundEnd,      0.0f, 0.5f, 1.0f, 1.0f);
```
In practice this is water, not underground (notice how the color is blue)

aurelienlt replied:
updated

sprunk commented on rts/Lua/LuaUnsyncedCtrl.cpp:
```diff
@@ -3382,6 +3384,226 @@ int LuaUnsyncedCtrl::SetBuildFacing(lua_State* L)
+}
+
+
+static bool ParseBuildSquareOptions(
```
I'd remove the explicit color arguments now that cmdColors for build squares exist.

aurelienlt replied:
done

sprunk commented on rts/Lua/LuaUnsyncedCtrl.cpp:
```diff
@@ -3382,6 +3384,226 @@ int LuaUnsyncedCtrl::SetBuildFacing(lua_State* L)
+		}
+	}
+
+	for (BuildInfo &buildInfo: buildInfos) {
+		unitDrawer->AddLuaBuildSquare(buildInfo, opts);
```
If this is just a trivial loop then I'd get rid of the function, one can iterate in Lua.

aurelienlt replied:
removed

sprunk commented on rts/Rendering/Units/UnitDrawer.cpp:
```diff
@@ -1333,6 +1329,182 @@ bool CUnitDrawerLegacy::ShowUnitBuildSquare(const BuildInfo& buildInfo, const st
 }
 
+static inline bool IsZeroColor(SColor color) {
+	return color.r == 0 && color.g == 0 && color.b == 0 && color.a == 0;
```
Aren't these floats (i.e. `0.0f` rather than `0`)?

aurelienlt replied:
removed

sprunk commented on rts/Rendering/Units/UnitDrawer.cpp:
```diff
@@ -1333,6 +1329,182 @@ bool CUnitDrawerLegacy::ShowUnitBuildSquare(const BuildInfo& buildInfo, const st
+	SColor startColor(cmdColors.undergroundStart);
+	SColor endColor(cmdColors.undergroundEnd);
+
+	for (int i=0; i<luaBuildPos.size(); i+=2) {
+		float3 pos1 = luaBuildPos[i];
+		float3 pos2 = luaBuildPos[i+1];
```
A code comment on why you're consuming 2 at a time and what happens if the number of entries in the table is not a multiple of 2 would be good

aurelienlt replied:
commented

sprunk commented on rts/Rendering/Units/UnitDrawer.cpp:
```diff
@@ -1333,6 +1329,182 @@ bool CUnitDrawerLegacy::ShowUnitBuildSquare(const BuildInfo& buildInfo, const st
+	return color.r == 0 && color.g == 0 && color.b == 0 && color.a == 0;
+}
+
+void CUnitDrawerLegacy::AddLuaBuildSquare(const BuildInfo& buildInfo, LuaBuildSquareOptions& opts) {
```
Could add a FIXME code comment that this is partially a copypaste of function X from file Y.cpp done due to compilation error Z

aurelienlt replied:
I think you meant `ParseBuildInfo`, it has been commented

sprunk commented on rts/Lua/LuaUnsyncedCtrl.cpp:
```diff
@@ -3382,6 +3384,226 @@ int LuaUnsyncedCtrl::SetBuildFacing(lua_State* L)
+}
+
+
+/***
+ *
```
A description on where to use the function would be good. Something like "best used in wupget:DrawFoo since that is the phase engine draws it natively, also works in DrawBaz and DrawQuux"

aurelienlt replied:
done

sprunk commented on rts/Lua/LuaUnsyncedCtrl.cpp:
```diff
@@ -3382,6 +3384,226 @@ int LuaUnsyncedCtrl::SetBuildFacing(lua_State* L)
+
+	int unitDefID = lua_toint(L, idx);
+	float3 pos = {lua_tofloat(L, idx+1), lua_tofloat(L, idx+2), lua_tofloat(L, idx+3)};
+	int facing = lua_toint(L, idx+4);
```
IIRC there's some sort of facing parser in LuaUtils that makes facing args work both with numbers and things like "south", might want to optionally check it out

aurelienlt replied:
No, all functions using facing (including `SetBuildFacing`) just parse an int.

lhog commented:
Looks like `DrawBuildSquare` is supposed to be called every draw frame. Is it really the best way to to do it? `AddLuaBuildSquare()` does quite a bit of calculations, though put into the cache. Won't it be better to introduce some sort of state into that call, e.g.
1) AddDrawBuildSquare/DelDrawBuildSquare
2) encode how often you want to refresh if the state is current per DrawBuildSquare, the default will be at 1 synced frame.
3) SetLuaDrawBuildColors() to control the colorization
4) (Maybe) for folks like ZK it will be nice to also allow for supplying the desired abs or relative height values of the grid or height function similar to how Spring.SetHeightMapFunc() operates.

aurelienlt commented:
This is actually the behavior of the original function `ShowUnitBuildSquare`: being called each rendering. Unless `Update` is called more often than `Draw`, I'm not fully clear about the enegine, but it feels like they are called as much, except when speeding a replay, in which case there is no square to show. Maybe there is a better moment to call `DrawBuildSquare` than in `Update`, it just seemed convenient since this is where building positions would be calculated by lua (this is the case for templates for instaces).

The squares are cached, so having Add/Del would complexify the behavior quite much for a limited gain. Especially since one player would only see HIS squares, only while holding the mouse button, and they would (almost) always be on screen.

About `SetLuaDrawBuildColors`, I'm not sure what would be the behavior. If the point is to customize some colors, I think it would make more sense to have `SetCmdColors(colorsMap) -> originalColorsMap` call that would let lua replace any color listed in `CCommandColors::colorNames`.

About custom height, I'm not sure what would be the use case. The current building command resolves height automatically.

lhog commented:
`ShowUnitBuildSquare` is not a great example as it updates the state and draws at the same time. One evidence is that ugly cache vector that was introduced some time ago to make it work a bit faster. In fact the state should be updated only once per synced frame (perhaps even less frequently) or whenever the mouse moves.

> The squares are cached, so having Add/Del would complexify the behavior quite much for a limited gain. Especially since one player would only see HIS squares, only while holding the mouse button, and they would (almost) always be on screen.

It will introduce some flexibility as in time we can figure out how to update the state smarter than just brute-forcing TestUnitBuildSquare() on every opportunity. For now at least users will be able to control how frequently the state should get brute-forced, unlike how it's done currently in the PR, where it's fixed to be [CACHE_VALIDITY_PERIOD](https://github.com/beyond-all-reason/spring/pull/1733/files#diff-e5733cd172ae4abf67fe875d3c3a134aab2689232f077baae8dbf0ba527bea02R1356)

Disregard my `SetLuaDrawBuildColors` comment, I misread how it's implemented in your PR.

> About custom height, I'm not sure what would be the use case. The current building command resolves height automatically.

ZK has a concept of terrain terraforming, someone might want to draw the grid at the future height of X or even with some custom height shape.

sprunk commented:
> (Maybe) for folks like ZK it will be nice to also allow for supplying the desired abs or relative height values of the grid or height function similar to how Spring.SetHeightMapFunc() operates.

The current interface in the PR is already good because it lets you specify position as XYZ. So it would be a matter of the renderer not ignoring the Y. Perhaps if that happens then Y could become optional so that specifying `nil` makes it use the unsynced ground height.

aurelienlt commented:
OK I understand the idea of Add/Del, it would permit for later optimisation. Looking at `TestBuildSquare` I'm not exactly sure what make it slow, seems like most steps are just about looking for values in 2D arrays. I suppose the best move would be to run this in a seperate thread, so it would draw stuffs once it has finished computed values, while displaying the old values when it takes longer than a frame.

For terraforming, unfortunately `ShowUnitBuildSquare` cannot really apply because I imagine that terraforming isn't associated to a `unitDefId`. It's more about adding the possibility to display arbitrary squares with arbitrary color.

About skipping the `GetBuildHeight` step, I can remove it from `AddLuaBuildSquare` since lua is already using `GetBuildHeight`, I'll have to check little more to know if I can remove it from `TestUnitBuildSquare` because it is called in quite many places.

aurelienlt commented:
I've checked how this is implemented in VAO:

- all submiussions are store in an array
- when adding, it returns the added index
- when removing, it replace the element by the tail
- lua has to make a complicated dance of maintaining the right index to know how to remove

So do we want to mimic that implementation, or we should simplify using a map ? I feel like using a map is better because we can trust C++ to do a faster job at updating a map than lua is doing.

aurelienlt commented:
OK next iteration. Now have `AddBuildSquare` and `RemoveBuildSquare`. Everything is stored in an `unordered_map` that also serves as cache (elements are kept few frames more). Calls to `TestUnitBuildSquare` are performed in the draw phase. That should make spamming `AddBuildSquare` much less dangerous because because it will only prduce map access until the draw phase. And obviously it will make it possible to optimize calls to only add / remove what is required.

lhog commented:
> I'm not exactly sure what make it slow, seems like most steps are just about looking for values in 2D arrays.

By itself is not that expensive (although it's much more than just a lookup in 2D array), but when N constructors queue M buildings it's NxM runs to perform and it becomes very expensive.

> I suppose the best move would be to run this in a seperate thread, so it would draw stuffs once it has finished computed values, while displaying the old values when it takes longer than a frame.

This is in line with what I'm saying. Even if implemented naively, the only place something might change is at game frame time. The game runs 30 sim frames per seconds, while rendering frames normally reach 3 digit figures.

> For terraforming, unfortunately ShowUnitBuildSquare cannot really apply because I imagine that terraforming isn't associated to a unitDefId. It's more about adding the possibility to display arbitrary squares with arbitrary color.

Terraforming defines the final state of heightmap, but it's ok if you skip this one.

> I've checked how this is implemented in VAO

Unsure how VAO relates to the topic of discussion. But that actually led me to some interesting thought. The way we store the buildable/unbuildable/feature squares is very expensive and redundant. It's enough to store the build-ability data in a 2D value array, possibly a bitmap and use shader to colorize the output according to the chosen colors.
I may try to implement it today for `ShowUnitBuildSquare()`.

lhog commented:
> OK next iteration. Now have `AddBuildSquare` and `RemoveBuildSquare`. Everything is stored in an `unordered_map` that also serves as cache (elements are kept few frames more). Calls to `TestUnitBuildSquare` are performed in the draw phase. That should make spamming `AddBuildSquare` much less dangerous because because it will only prduce map access until the draw phase. And obviously it will make it possible to optimize calls to only add / remove what is required.

Ok, I'll have a look today.

lhog commented on rts/Rendering/Units/UnitDrawerCache.h:
```diff
@@ -0,0 +1,45 @@
+	bool operator()(const LuaBuildSquareTaskKey& buildKey1, const LuaBuildSquareTaskKey& buildKey2) const;
+};
+
+using LuaBuildSquareTasksMap = std::unordered_map<LuaBuildSquareTaskKey, LuaBuildSquareTask, LuaBuildSquareTaskHash, LuaBuildSquareTaskHash>;
```
spring::unordered_map is supposed to perform better.

aurelienlt replied:
Oh good ill check tomorrow

aurelienlt replied:
done

lhog commented on rts/Lua/LuaUnsyncedCtrl.cpp:
```diff
@@ -3382,6 +3384,93 @@ int LuaUnsyncedCtrl::SetBuildFacing(lua_State* L)
+	BuildInfo buildInfo = ParseBuildInfo(L, __func__, 1);
+	LuaBuildSquareOptions opts = ParseBuildSquareOptions(L, __func__, 6);
+
+	unitDrawer->AddLuaBuildSquare(buildInfo, opts);
```
Add time to cache too.

aurelienlt replied:
I'll add that in the options

aurelienlt replied:
added `cacheValidity` option

aurelienlt commented:
> The way we store the buildable/unbuildable/feature squares is very expensive and redundant. It's enough to store the build-ability data in a 2D value array, possibly a bitmap and use shader to colorize the output according to the chosen colors.
I may try to implement it today for ShowUnitBuildSquare().

Oh yeah I absolutely hate how the test function takes 3 vector pointers, instead of just returning a struct with a 2D array of states.

> The game runs 30 sim frames per seconds, while rendering frames normally reach 3 digit figures.

I've been trying to understand how threading works, but I didn't get much clues through the code. I just noticed that despite having a very decent laptop, replay speed barely exceeds normal speed on big battles. I suppose this is the cause of unsynchronisations that players seem to complain about.

lhog commented on rts/Lua/LuaUnsyncedCtrl.cpp:
```diff
@@ -171,6 +171,8 @@ bool LuaUnsyncedCtrl::PushEntries(lua_State* L)
 
 	REGISTER_LUA_CFUNC(DrawUnitCommands);
+	REGISTER_LUA_CFUNC(AddBuildSquare);
+	REGISTER_LUA_CFUNC(RemoveBuildSquare);
```
I think it's a good foundation from the API cleanness perspective, but perhaps add `Unit` in between `{Add,Remove}` and `Build`?

aurelienlt replied:
done

sprunk commented on rts/Game/GameHelper.cpp:
```diff
@@ -1283,6 +1283,7 @@ CGameHelper::BuildSquareStatus CGameHelper::TestUnitBuildSquare(
 	// const float buildHeight = GetBuildHeight(testPos, buildInfo.def, synced);
 	// const float modelHeight = (model != nullptr) ? math::fabs(model->height) : 10.0f;
 
+	// FIXME this call is probably useless, we can assume that any caller already assigned it correctly
```
can we? Lua can pass anything here.

aurelienlt replied:
> The current interface in the PR is already good because it lets you specify position as XYZ. So it would be a matter of the renderer not ignoring the Y. Perhaps if that happens then Y could become optional so that specifying nil makes it use the unsynced ground height.

That was following following this comment, to read provided Y instead of recomputing it. I don't think we need to worry about LUA passing stupid values, they can call `GetBuildHeight` too. But that was following your comment, so your call to keep this FIXME around or not.

sprunk replied:
I meant more in the wording sense. Games decide what is or isn't "correct", not engine, and it wasn't obvious what the practical effect of removal would be. I'd keep the line below it for now until it becomes possible to specify "use ground height" in addition to an explicit value for Y (because the ground height can change and it would otherwise be a pain to adjust existing squares) and just change the comment to be more explicit
```suggestion
	// FIXME remove this line to let Lua buildsquares specify arbitrary Y, if/when defaulting to ground height exists
```

lhog committed:
Merge branch 'master' into feature/lua-show-buid-square

Commit:
Feature: register new lua functions `AddBuildSquare` and `RemoveBuildSquare` to show build squares on custom building commands

aurelienlt committed:
Reword comment

Co-authored-by: sprunk <spr.ng@o2.pl>

aurelienlt commented:
Hi, anything else you want me to improve, or we're good to merge ?

sprunk commented on rts/Lua/LuaUnsyncedCtrl.cpp:
```diff
@@ -3541,6 +3543,91 @@ int LuaUnsyncedCtrl::SetBuildFacing(lua_State* L)
+		luaL_checkfloat(L, idx+2),
+		luaL_checkfloat(L, idx+3),
+	};
+	int facing = luaL_checkinteger(L, idx+4);
```
Use ParseFacing
https://github.com/beyond-all-reason/spring/blob/38a5deda9d42c14509adcadee054163f7cf0adb8/rts/Lua/LuaUtils.cpp#L1170

aurelienlt replied:
done

sprunk commented on rts/Lua/LuaUnsyncedCtrl.cpp:
```diff
@@ -3541,6 +3543,91 @@ int LuaUnsyncedCtrl::SetBuildFacing(lua_State* L)
+ * @number y
+ * @number z
+ * @number facing
+ * @bool[opt] unbuildable
```
Seems to be a table now.

aurelienlt replied:
I've tried to fix those comments, I'm still not clear how they work though because they are quite different from LUA annotations.

sprunk commented on rts/Lua/LuaUnsyncedCtrl.cpp:
```diff
@@ -3541,6 +3543,91 @@ int LuaUnsyncedCtrl::SetBuildFacing(lua_State* L)
+	const char* caller,
+	int idx
+) {
+	int unitDefID = luaL_checkinteger(L, idx);
```
Should error if invalid unitDefID.

aurelienlt replied:
done

sprunk reviewed (commented):
Documentation would be good:
* what happens if you call Add twice with the same args, or if the Y is different? How lenient is duplicate detection?
* do x/z need to be aligned to the grid?
* what is "cache validity"? Just the rate of buildability checks or will the square remove itself?
* does Remove require 100% identical args that were passed to Add to successfully remove? If not, how much leniency is there?

sprunk commented:
Some widget so we could test the main use case would be decent too

aurelienlt commented:
Hi I created a small widget to test the new call:
https://github.com/aurelienlt/Beyond-All-Reason/blob/debug/test-build/luaui/Widgets/test_build.lua
By pressing "B" when a build command is selected, you switch to the widget, which let your dump some buildings, similarily to my GIF. This way also can see that the normal build mode still works fine.

Commit:
Feature: register new lua functions `AddBuildSquare` and `RemoveBuildSquare` to show build squares on custom building commands

aurelienlt commented:
I've added some checks and fixes to avoid crashing with unexpected arguments. Surprisingly, I've found out that the engine doesn't enforce a building to follow the grid, so `GiveOrderToUnit` let your build in weird places. I haven't added code to enforce a valid building position, during my tests seems like grid is snapped to the right place though, probably because of `int(testPos.x / SQUARE_SIZE)` in `CGameHelper::BuildSquareStatus`.

aurelienlt commented:
@sprunk So the current behavior is to never recompute the squares, and keep them in cache when removed according to `cacheValidity`. Mostly because I wanted to avoid having to recompute too many squares at once whereas they are only used for something very temporary, so I assumed players wouldn't be annoyed by corner cases.

I'm not sure what would be the best behavior, I'm open to suggestions. We could recompute them regularily once validity is passed despite being still used, we could add another argument for setting a refresh time ...

Also I've tested more thoroughly `cacheValidity` and fixed it.

sprunk reviewed (commented):
> So the current behavior is to never recompute the squares, and keep them in cache when removed according to cacheValidity.

I thought the whole point of `cacheValidity` was to be the recomputation period, there's little point keeping squares in cache otherwise. Generally, it should behave the same as when engine draws the squares for regular line build in this regard.

There could be some way to get rid of buildsquares you've lost handles to. This means either a "get squares in area" or some sort of "clear all" function.
